lib/elixir/lib/enum.ex

```diff
@@ -3236,21 +3242,31 @@ defmodule Enum do
       iex> Enum.with_index([:a, :b, :c], 3)
       [a: 3, b: 4, c: 5]
 
+      iex> Enum.with_index([:a, :b, :c], fn element, index -> {index, element} end)
+      [{0, :a}, {1, :b}, {2, :c}]
   """
-  @spec with_index(t, integer) :: [{element, index}]
-  def with_index(enumerable, offset \\ 0) do
+  @spec with_index(t, integer | (element, index -> any)) :: list
```

Comment:
```suggestion
  @spec with_index(t, integer) :: [{term, integer}]
  @spec with_index(t, (element, index -> value)) :: [value] when value: any
```

Reply:
Didn't even knew you could write specs this way T_T